zapr_test.go

```diff
@@ -0,0 +1,376 @@
+			format: `{"ts":%f,"caller":"zapr/zapr_test.go:%d","msg":"duration value argument","v":0,"duration":"5s"}
+`,
+			keysValues: []interface{}{"duration", time.Duration(5 * time.Second)},
+		},
+	}
+
+	test := func(t *testing.T, logNumeric *string, enablePanics *bool, allowZapFields *bool, data testCase) {
+		var buffer bytes.Buffer
+		writer := bufio.NewWriter(&buffer)
+		var sampleInfoLogger logr.Logger
+		zl := newZapLogger(zapcore.Level(-100), zapcore.AddSync(writer))
+		if logNumeric == nil && enablePanics == nil && allowZapFields == nil {
+			// No options.
+			sampleInfoLogger = zapr.NewLogger(zl)
+		} else {
+			opts := []zapr.Option{}
+			if logNumeric != nil {
+				opts = append(opts, zapr.LogInfoLevel(*logNumeric))
+			}
+			if enablePanics != nil {
+				opts = append(opts, zapr.PanicMessages(*enablePanics))
+			}
+			if allowZapFields != nil {
+				opts = append(opts, zapr.AllowZapFields(*allowZapFields))
+			}
+			sampleInfoLogger = zapr.NewLoggerWithOptions(zl, opts...)
+		}
+		if data.withKeysValues != nil {
+			sampleInfoLogger = sampleInfoLogger.WithValues(data.withKeysValues...)
+		}
+		for _, name := range data.names {
+			sampleInfoLogger = sampleInfoLogger.WithName(name)
+		}
+		if data.wrapper != nil {
+			data.wrapper(sampleInfoLogger, data.msg, data.keysValues...)
+		} else {
+			sampleInfoLogger.Info(data.msg, data.keysValues...)
+		}
+		writer.Flush()
+		logStr := buffer.String()
+
+		logStrLines := strings.Split(logStr, "\n")
+		var dataFormatLines []string
+		noPanics := enablePanics != nil && !*enablePanics
+		withZapFields := allowZapFields != nil && *allowZapFields
+		for _, line := range strings.Split(data.format, "\n") {
+			// Potentially filter out all or some panic
+			// message. We can recognize them based on the
+			// expected special keys.
+			if strings.Contains(line, "invalid key") ||
+				strings.Contains(line, "ignored key") {
+				if noPanics {
+					continue
+				}
+			} else if strings.Contains(line, "zap field") {
+				if noPanics || withZapFields {
+					continue
+				}
+			}
+			haveZapField := strings.Index(line, `"zap-field`)
+			if haveZapField != -1 && !withZapFields && !strings.Contains(line, "zap field") {
+				// When Zap fields are not allowd, output gets truncapted at the first Zap field.
+				line = line[0:haveZapField-1] + "}"
+			}
+			dataFormatLines = append(dataFormatLines, line)
+		}
+		if !assert.Equal(t, len(logStrLines), len(dataFormatLines)) {
+			t.Errorf("Info has wrong format: no. of lines in log is incorrect \n expected: %s\n got: %s", dataFormatLines, logStrLines)
+		}
+
+		for i := range logStrLines {
+			if len(logStrLines[i]) == 0 && len(dataFormatLines[i]) == 0 {
+				continue
+			}
+			var ts float64
+			var lineNo int
+			format := dataFormatLines[i]
+			if logNumeric == nil || *logNumeric == "" {
+				format = regexp.MustCompile(`,"v":-?\d`).ReplaceAllString(format, "")
+			}
+			n, err := fmt.Sscanf(logStrLines[i], format, &ts, &lineNo)
```

Comment:
Why `scanf` rather than `json.Unmarshal` ?

Reply:
That was copied. My guess is that it was used because unmarshalling would need different structs for each test case whereas a format string works for all of them as long as it has `%f` and `%d` embedded somewhere.